https://github.com/canonical/cloud-init/commit/8ec8c3fc63a59b85888a0b52356b784314a1d4cc

=== Begin SRU Template ===
[Impact]

Azure's Mellanox interfaces are presented to the instance with identical MAC
addresses to their corresponding non-Mellanox interface: cloud-init treats
interfaces with duplicate MACs as an error condition (because it doesn't know
which interface to apply network configuration to).  This causes cloud-init to
fail to apply network configuration; as this happens on every boot on Azure,
this means that the interfaces are not readily usable.

[Test Case]

1) Launch a Standard_HC44rs instance on Azure
2) Install `rdma-core` (so the Mellanox interfaces are configured)
3) Reboot
4) Observe the buggy behaviour (warning in log)
5) Install the -proposed cloud-init
6) Reboot
7) Observe the correct behaviour (no additional warning in log)

(This issue does not reproduce on xenial.)

[Regression Potential]

This introduces generic support in cloud-init's networking code for excluding
interfaces using a particular driver from the interfaces that cloud-init will
consider.  This support is used only by `DataSourceAzure`, which blacklists
only Mellanox drivers.

(After landing upstream, we identified a (latent) upgrade bug and addressed
that with https://github.com/canonical/cloud-init/pull/659.)

[Test Output]

(% indicates a local prompt, $ a remote one.)

== bionic ==

% az vm create -n bionic-$(date +%s) -g xx16resourcegroup --image Canonical:UbuntuServer:18.04-LTS:18.04.202012010 --ssh-key-value /home/daniel/.ssh/id_rsa.canonical.pub -l westus2 --admin-username ubuntu --size Standard_HC44rs                     [ 16:34:10 ]
{
  "fqdns": "",
  "id": "/subscriptions/9a37cc2c-dc4a-4097-84dd-45dd2b8cbc63/resourceGroups/xx16resourcegroup/providers/Microsoft.Compute/virtualMachines/bionic-1607636069",
  "location": "westus2",
  "macAddress": "00-0D-3A-C6-7A-30",
  "powerState": "VM running",
  "privateIpAddress": "10.0.0.7",
  "publicIpAddress": "52.247.252.59",
  "resourceGroup": "xx16resourcegroup",
  "zones": ""
}
% ssh ubuntu@52.247.252.59
$ sudo apt update
...
$ sudo apt install rdma-core
...
Setting up rdma-core (17.1-1ubuntu0.2) ...
...
$ sudo reboot
% ssh ubuntu@52.247.252.59
$ grep WARN /var/log/cloud-init.log 
2020-12-10 22:16:46,255 - util.py[WARNING]: failed stage init-local
$ echo deb "http://archive.ubuntu.com/ubuntu $(lsb_release -sc)-proposed main" | sudo tee /etc/apt/sources.list.d/proposed.list
deb http://archive.ubuntu.com/ubuntu bionic-proposed main
$ sudo apt update
...
$ sudo apt install cloud-init
..
Setting up cloud-init (20.4-0ubuntu1~18.04.1) ...
...
$ sudo reboot
% ssh ubuntu@52.247.252.59
$ grep WARN /var/log/cloud-init.log  # Only the warning from before
2020-12-10 22:16:46,255 - util.py[WARNING]: failed stage init-local

== focal ==

% az vm create -n focal-$(date +%s) -g xx16resourcegroup --image Canonical:0001-com-ubuntu-server-focal:20_04-lts:20.04.202012010 --ssh-key-value /home/daniel/.ssh/id_rsa.canonical.pub -l westus2 --admin-username ubuntu --size Standard_HC44rs
{
  "fqdns": "",
  "id": "/subscriptions/9a37cc2c-dc4a-4097-84dd-45dd2b8cbc63/resourceGroups/xx16resourcegroup/providers/Microsoft.Compute/virtualMachines/focal-1607635853",
  "location": "westus2",
  "macAddress": "00-0D-3A-FB-7A-74",
  "powerState": "VM running",
  "privateIpAddress": "10.0.0.5",
  "publicIpAddress": "20.191.96.220",
  "resourceGroup": "xx16resourcegroup",
  "zones": ""
}
% ssh ubuntu@20.191.96.220
$ sudo apt update
...
$ sudo apt install rdma-core
...
Setting up libnl-3-200:amd64 (3.4.0-1) ...
Setting up rdma-core (28.0-1ubuntu1) ...
...
$ sudo reboot
% ssh ubuntu@20.191.96.220
$ grep WARN /var/log/cloud-init.log
2020-12-10 21:56:47,726 - util.py[WARNING]: failed stage init-local
$ echo deb "http://archive.ubuntu.com/ubuntu $(lsb_release -sc)-proposed main" | sudo tee /etc/apt/sources.list.d/proposed.list
deb http://archive.ubuntu.com/ubuntu focal-proposed main
$ sudo apt update
...
$ sudo apt install cloud-init
...
Setting up cloud-init (20.4-0ubuntu1~20.04.1) ...
...
$ sudo reboot
% ssh ubuntu@20.191.96.220
$ grep WARN /var/log/cloud-init.log  # Only the warning from before
2020-12-10 21:56:47,726 - util.py[WARNING]: failed stage init-local

== groovy ==

% az vm create -n groovy-$(date +%s) -g xx16resourcegroup --image Canonical:0001-com-ubuntu-server-groovy:20_10:20.10.202012070 --ssh-key-value /home/daniel/.ssh/id_rsa.canonical.pub -l westus2 --admin-username ubuntu --size Standard_HC44rs
{
  "fqdns": "",
  "id": "/subscriptions/9a37cc2c-dc4a-4097-84dd-45dd2b8cbc63/resourceGroups/xx16resourcegroup/providers/Microsoft.Compute/virtualMachines/groovy-1607635876",
  "location": "westus2",
  "macAddress": "00-0D-3A-FC-45-27",
  "powerState": "VM running",
  "privateIpAddress": "10.0.0.6",
  "publicIpAddress": "20.190.38.200",
  "resourceGroup": "xx16resourcegroup",
  "zones": ""
}
% ssh ubuntu@20.190.38.200
$ sudo apt update
...
$ sudo apt install rdma-core
...
Setting up libnl-3-200:amd64 (3.4.0-1) ...
Setting up rdma-core (31.0-1) ...
...
$ sudo reboot
% ssh ubuntu@20.190.38.200
$ grep WARN /var/log/cloud-init.log
2020-12-10 22:16:38,150 - util.py[WARNING]: failed stage init-local
$ echo deb "http://archive.ubuntu.com/ubuntu $(lsb_release -sc)-proposed main" | sudo tee /etc/apt/sources.list.d/proposed.list
deb http://archive.ubuntu.com/ubuntu groovy-proposed main
$ sudo apt update
...
$ sudo apt install cloud-init
...
Setting up cloud-init (20.4-0ubuntu1~20.10.1) ...
...
$ sudo reboot
% ssh ubuntu@20.190.38.200
$ grep WARN /var/log/cloud-init.log  # Only the warning from before
2020-12-10 22:16:38,150 - util.py[WARNING]: failed stage init-local
